src/assert_not_impl.rs

```diff
@@ -0,0 +1,107 @@
+/// Asserts that the type does **not** implement all of the given traits.
+///
+/// This can be used to ensure types implement auto traits such as [`Send`] and
+/// [`Sync`], as well as traits with [blanket `impl`s][blanket].
+///
+/// # Examples
+///
+/// On stable Rust, using the macro requires a unique “label” when used in a
+/// module scope:
+///
+#[cfg_attr(feature = "nightly", doc = "```ignore")]
+#[cfg_attr(not(feature = "nightly"), doc = "```")]
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {}
+/// assert_not_impl!(str; *const u8, Sync, From<&'static str>);
+/// assert_not_impl!(vec; *const u8, Sync);
+/// ```
+///
+/// The [labeling limitation](index.html#limitations) is not necessary if
+/// compiling on nightly Rust with the `nightly` feature enabled:
+///
+#[cfg_attr(feature = "nightly", doc = "```")]
+#[cfg_attr(not(feature = "nightly"), doc = "```ignore")]
+/// #![feature(underscore_const_names)]
+/// # #[macro_use] extern crate static_assertions;
+///
+/// assert_not_impl!(&'static mut u8, Copy);
+///
+/// fn main() {
+///     assert_not_impl!(u32, Into<usize>);
+/// }
+/// ```
+///
+/// The following example fails to compile since `u32` can be converted into `u64`.
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # fn main() {
+/// assert_not_impl!(u32, Into<u64>);
+/// # }
+/// ```
+///
+/// Note that all traits given in a single invocation need to not be implemented. If you want to
+/// check that none of multiple trait are implemented you must invoke the macro several times.
+///
+#[cfg_attr(feature = "nightly", doc = "```ignore")]
+#[cfg_attr(not(feature = "nightly"), doc = "```")]
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // Cell<u32> is not both Sync and Send
+/// assert_not_impl!(cell; Cell<u32>, Sync, Send);
+/// ```
+///
+/// ```compile_fail
+/// # #[macro_use] extern crate static_assertions;
+/// # use static_assertions::_core::cell::Cell;
+/// # fn main() {}
+/// // But it is Send, this fails to compile
+/// assert_not_impl!(cell; Cell<u32>, Send);
+/// ```
+///
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [`Sync`]: https://doc.rust-lang.org/std/marker/trait.Sync.html
+/// [blanket]: https://doc.rust-lang.org/book/second-edition/ch10-02-traits.html#using-trait-bounds-to-conditionally-implement-methods
+#[macro_export(local_inner_macros)]
+macro_rules! assert_not_impl {
+    ($($xs:tt)+) => { _assert_not_impl!($($xs)+); };
+}
+
+#[doc(hidden)]
+#[cfg(feature = "nightly")]
+#[macro_export(local_inner_macros)]
+macro_rules! _assert_not_impl {
+    ($x:ty, $($t:path),+ $(,)*) => {
+        const _: fn() -> () = || {
+            struct Check<T: ?Sized>(T);
+            trait AmbiguousIfImpl<A> { fn some_item() { } }
+
+            impl<T: ?Sized> AmbiguousIfImpl<()> for Check<T> { }
+            impl<T: ?Sized $(+ $t)*> AmbiguousIfImpl<u8> for Check<T> { }
```

Comment:
Rather than use `u8` here, add `struct Invalid;` under `Check` (or a better name if you think of one). Since it never gets constructed, make sure to add `#[allow(dead_code)]` to the declaration.

For `assert_not_impl_any!`, you should implement this bit like the following. Notice that each `Invalid` type is unique because of the scoping.
```rust
$({
    #[allow(dead_code)]
    struct Invalid;

    impl<A: ?Sized + $t> AmbiguousIfImpl<Invalid> for Check<A> {}
})+
```

Also, you should be matching multiple `$t` via `+` and not `*`.